phygnn/layers/custom_layers.py

```diff
@@ -1344,20 +1423,119 @@ def call(x, hi_res_feature):
         hi_res_feature : tf.Tensor | np.ndarray
             This should be a 4D array for spatial enhancement model or 5D array
             for a spatiotemporal enhancement model (obs, spatial_1, spatial_2,
-            (temporal), 1) that can be used to fix values of x. This is NaN
-            where there are no observations and real values where observations
-            exist.
+            (temporal), 1). This is NaN where there are no observations and
+            real values where observations exist.
 
         Returns
         -------
         x : tf.Tensor
             Output tensor with the hi_res_feature used to fix values of x.
         """
+        if hi_res_feature is None:
+            hi_res_feature = tf.fill(x[..., :1].shape, np.nan)
         mask = tf.math.is_nan(hi_res_feature)
         fixed = tf.where(mask, x[..., :1], hi_res_feature)
         return tf.concat((x, fixed), axis=-1)
 
 
+class Sup3rConcatEmbeddedObs(tf.keras.layers.Layer):
+    """Layer to concatenate sparse data in the middle of a super resolution
+    forward pass, with a learned embedding. This is used to condition models
+    on sparse observation data."""
+
+    def __init__(self, filters=32, name=None):
+        """
+        Parameters
+        ----------
+        filters : int
+            Number of filters to use for each convolutional layer. e.g.
+            [64, 32] for one layer with 64 and another with 32.
+        name : str | None
+            Unique str identifier of the layer. Usually the name of the
+            hi-resolution feature used in the concatenation.
+        """
+        super().__init__(name=name)
+        self.embed_net = None
+        self.rank = None
+        self.filters = filters if isinstance(filters, list) else [filters]
+
+    def build(self, input_shape):
+        """Build the weight net layer based on an input shape
+
+        Parameters
+        ----------
+        input_shape : tuple
+            Shape tuple of the input tensor
+        """
+        self.rank = len(input_shape)
+
+        if self.rank == 4:
+            conv_class = tf.keras.layers.Conv2D
+        else:
+            conv_class = tf.keras.layers.Conv3D
+        self.embed_net = []
+        for _, filters in enumerate(self.filters):
+            self.embed_net += [
+                conv_class(
+                    filters=filters,
+                    kernel_size=1,
```

Comment:
kernel size 1? Isn't a conv with kernel 1 just a linear transform (scalar and adder)? If this is the case, no need for multiple?

Where are the weights for this embed network actually trained? I don't think the optimizer will see these weights.

Reply:
Theres still mixing along channels, just done point wise. And non-linear mixing is done with multiple layers. This like applying multiple dense layers point wise. You scared me but the weights show up in layer.trainable_weights. Doesn't that automatically happen if the layers are defined in __init__ or build? We've defined hidden layers similarly elsewhere, no?

Reply:
Okay gotcha on the activations and similar to dense layers. 

Yeah i think it needs to be in `Sup3rConcatEmbeddedObs.trainable_weights`. I'm a little surprised that the weights from an attribute show up there but if you've confirmed then should be all good. Here's an example where i think `tf.keras.layers.Layer.add_weight()` makes sure the weight is trainable, i assume something in the Conv object init does the same thing and somehow has awareness that its an attribute in a trainable layer?

https://github.com/NREL/phygnn/blob/main/phygnn/layers/custom_layers.py#L209-L215

Reply:
Yeah I saw that you did that with the Gaussian layer. I think when the model is initialized the layers are added to the compute graph and since conv layers already have trainable weights this works out. I think that's sort of what we assumed with these other layers, like SqueezeAndExcitation with the _hidden_layers attrs.

Reply:
Yeah gotcha. I never actually used SqueezeAndExcitation but makes sense